Add ways to pass Analytics Host Entitlement info for out of process integrators

QilongTang opened the pull request:
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

- [x] Add analytics host entitlement info including parent id and session id. See linked PR: https://git.autodesk.com/Dynamo/Analytics.NET/pull/34
- [x] Add ability to pass in these two field using CLI or CLIWPF
- [x] Added unit tests 

These are done so that out of progress integrators like FormIt or Alias could pass in analytics info.

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@DynamoDS/dynamo 

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

QilongTang commented on src/DynamoCore/Updates/UpdateManager.cs:
```diff
@@ -164,6 +165,11 @@ public interface IUpdateManager
+        /// <summary>
+        /// Analytics info of the host
+        /// </summary>
+        HostContextInfo HostInfo { get; set; }
```
I am exposing this type defined in Analytics.Net here as we did for ProductInfo and CustomerInfo. The only difference is that `HostContextInfo` cant be constructed run time but rely on integrators to pass in, in that case, we will need this type in a couple of other places where I had to add reference.

QilongTang replied:
On another though.. This seems problematic in a way that integrators will now depend on Analytics.Net where this type is defined.. @mjkkirschner @pinzart This seems both API breaking and problematic.. Maybe we only expose two strings `ParentId` and `SessionId` and ask DynamoCore to construct them into `HostContextInfo`. What do you think?

mjkkirschner replied:
yeah, there seems to be 3 problems
* why is this property being added to IUpdateManager - what does this have to do with updates?
* this is breaking a public interface
* new dependency on analytics.net 


I'm fine with your proposed solution I think - not sure exactly what it would look like though - where would DynamoCore get those strings?

QilongTang replied:
@mjkkirschner @pinzart I decided to use the alternative approach on this. By inspecting https://git.autodesk.com/Alias/dynamo-plugin/blob/5324734fa78953c7575715733e59f1952b379d7b/src/DynamoIPC/UpdateManager.cs#L87, I think introducing this type at this level will not only break integration but also introduce extra dependency there

QilongTang replied:
@mjkkirschner These strings will be empty for Sandbox but passed in by integrators using CLI or when they define update manager. I think another good place might be `StartConfiguration` but given `HostName` is also in UpdateManager, for now I would like they all live in the same place

QilongTang commented on src/DynamoCore/Logging/DynamoAnalyticsClient.cs:
```diff
@@ -178,9 +180,10 @@ public DynamoAnalyticsClient(DynamoModel dynamoModel)
-            string buildId = "", releaseId = "";
-            Version version;
-            if (Version.TryParse(dynamoModel.Version, out version))
+            hostInfo = new HostContextInfo() { ParentId = dynamoModel.UpdateManager.ParentId, SessionId = dynamoModel.UpdateManager.SessionId };
```
We only reference `HostContextInfo` here so Analytics.Net refrence is kept only for DynamoCore

QilongTang commented on src/DynamoCore/Logging/DynamoAnalyticsClient.cs:
```diff
@@ -232,7 +235,7 @@ public void Start()
 
                 //If not ReportingAnalytics, then set the idle time as infinite so idle state is not recorded.
-                Service.StartUp(product, new UserInfo(Session.UserId), TimeSpan.FromMinutes(30));
+                Service.StartUp(product, new UserInfo(Session.UserId), hostInfo, TimeSpan.FromMinutes(30));
```
Arguably we can add hostInfo as 4th param and make it optional but since only Dynamo will be consuming this API, I think this should be acceptable

QilongTang commented:
@DynamoDS/dynamo This PR is now ready for another look.

pinzart90 commented on src/DynamoApplications/StartupUtils.cs:
```diff
@@ -220,7 +233,21 @@ public static DynamoModel MakeModel(bool CLImode, string asmPath = "", string ho
+        /// <param name="parentId">Dynamo host parent id for analytics purpose</param>
+        /// <param name="sessionId">Dynamo host session id for analytics purpose</param>
+        /// <returns></returns>
+        public static DynamoModel MakeModel(bool CLImode, string asmPath = "", string hostName = "", string parentId = "", string sessionId = "")
```
maybe its time to put these analytics related properties in a separate struct ? might.. make it easier to pass around..

mjkkirschner replied:
yeah, if this is easy @QilongTang seems like a good idea to wrap them up into some options bag, we won't need to keep adding new constructors for MakeModel or worrying about binary compatibility with overloads.

QilongTang replied:
I have wrapped the strings into a struct as part of this PR

pinzart90 commented on src/DynamoApplications/StartupUtils.cs:
```diff
@@ -197,13 +206,17 @@ public static void PreloadShapeManager(ref string geometryFactoryPath, ref strin
         /// <returns></returns>
-        private static IUpdateManager InitializeUpdateManager(string hostName = "")
+        private static IUpdateManager InitializeUpdateManager(string hostName = "", String parentId = "", String sessionId = "")
         {
```
yeah, weird place to put them, UpdateManager
Can you add a separate jira task to refactor this (move them out) ?
Or maybe we can do it as part of this PR ?

QilongTang replied:
Will give that a try

pinzart90 commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -542,8 +542,7 @@ protected DynamoModel(IStartConfiguration config)
             IPreferences preferences = CreateOrLoadPreferences(config.Preferences);
-            var settings = preferences as PreferenceSettings;
-            if (settings != null)
+            if (preferences is PreferenceSettings settings)
```
I think this declaration pattern already checks for null value.
From msdn:
`Beginning with C# 7.0, a declaration pattern with type T matches an expression when an expression result is non-null and any of the following conditions are true:...`
Example:
https://docs.microsoft.com/en-us/dotnet/csharp/fundamentals/functional/pattern-matching#null-checks

QilongTang commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -431,6 +449,10 @@ public interface IStartConfiguration
+            /// <summary>
+            /// Host analytics info
+            /// </summary>
+            HostAnalyticsInfo HostAnalyticsInfo { get; set; }
```
This is now part of `IStartConfiguration` so integrators can set it there

mjkkirschner replied:
This is a breaking change - you cannot change a public interface like this - it means that anyone that implemented the interface will no longer conform to it. Instead add the new property to the concrete class DefaultStartConfiguration and if you prefer you can add a new interface.

QilongTang replied:
Thanks @mjkkirschner . Addressed

QilongTang commented:
@mjkkirschner @pinzart I cleaned up the solution and used a `HostAnalyticsInfo` struct in DynamoModel to hold pieces of info so this is scalable. Also it has been added to `IStartConfiguration` so integrators can set use the config or CLI. Let me know what you think

pinzart90 commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -183,6 +196,11 @@ public string Version
         /// </summary>
         public string HostName { get; set; }
```
maybe obsolete HostName ?

QilongTang replied:
Will do

pinzart90 commented on src/DynamoApplications/StartupUtils.cs:
```diff
@@ -196,14 +203,14 @@ public static void PreloadShapeManager(ref string geometryFactoryPath, ref strin
+        /// <param name="info">Dynamo host analytics info.</param>
         /// <returns></returns>
-        private static IUpdateManager InitializeUpdateManager(string hostName = "")
+        private static IUpdateManager InitializeUpdateManager(HostAnalyticsInfo info)
```
Do we still need InitializeUpdateManager(HostAnalyticsInfo) ?
or can we obsolete it and only used COnfig + DynamoModel ?

QilongTang replied:
We still need to initialize it but, we can remove the param. Since this is a private function, no need to obsolete

QilongTang commented:
Hi @pinzart @mjkkirschner Addressed all comments. I think FormIt should be covered with the CLI changes but still curious how Alias need to update their code base, if they stick to `IStartConfig` how do they set the `ParentId`?
